core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2015-2016 IBM Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.containerpool
+
+import scala.collection.mutable
+
+import akka.actor.Actor
+import akka.actor.ActorRef
+import akka.actor.ActorRefFactory
+import akka.actor.Props
+import whisk.common.AkkaLogging
+import whisk.core.dispatcher.ActivationFeed.ContainerReleased
+import whisk.core.entity.ByteSize
+import whisk.core.entity.CodeExec
+import whisk.core.entity.EntityName
+import whisk.core.entity.ExecutableWhiskAction
+import whisk.core.entity.size._
+
+sealed trait WorkerState
+case object Busy extends WorkerState
+case object Free extends WorkerState
+
+case class WorkerData(data: ContainerData, state: WorkerState)
+
+/**
+ * A pool managing containers to run actions on.
+ *
+ * This pool fulfills the other half of the ContainerProxy contract. Only
+ * one job (either Start or Run) is sent to a child-actor at any given
+ * time. The pool then waits for a response of that container, indicating
+ * the container is done with the job. Only then will the pool send another
+ * request to that container.
+ *
+ * Upon actor creation, the pool will start to prewarm containers according
+ * to the provided prewarmConfig, iff set. Those containers will **not** be
+ * part of the poolsize calculation, which is capped by the poolSize parameter.
+ * Prewarm containers are only used, if they have matching arguments
+ * (kind, memory) and there is space in the pool.
+ *
+ * @param childFactory method to create new containers
+ * @param poolSize maximum size of containers allowed in the pool
+ * @param feed actor to request more work from
+ * @param prewarmConfig optional settings for container prewarming
+ */
+class ContainerPool(
+    childFactory: ActorRefFactory => ActorRef,
+    poolSize: Int,
+    feed: ActorRef,
+    prewarmConfig: Option[PrewarmingConfig] = None) extends Actor {
+    val logging = new AkkaLogging(context.system.log)
+
+    val pool = new mutable.HashMap[ActorRef, WorkerData]
+    val prewarmedPool = new mutable.HashMap[ActorRef, WorkerData]
+
+    prewarmConfig.foreach { config =>
+        logging.info(this, s"pre-warming ${config.count} ${config.exec.kind} containers")
+        (1 to config.count).foreach { _ =>
+            prewarmContainer()
```

Comment:
The `foreach` in `prewarmContainer` is for the Option. Christian already commented on some weirdness here, i'll resolve.